haystack/nodes/query_classifier/transformers.py

```diff
@@ -57,43 +58,89 @@ class TransformersQueryClassifier(BaseQueryClassifier):
         :param use_gpu: Whether to use GPU (if available).
+        :param task: Specifies the type of classification. Possible values: 'text-classification' or 'zero-shot-classification'.
+        :param labels: If the task is 'text-classification' and an ordered list of labels is provided, the first label corresponds to output_1,
+        the second label to output_2, and so on. The labels must match the model labels; only the order can differ. Otherwise, model labels are considered.
+        If the task is 'zero-shot-classification', these are the candidate labels.
+        :param batch_size: The number of queries to be processed at a time.
         """
         super().__init__()
-
-        self.devices, _ = initialize_device_settings(use_cuda=use_gpu)
+        devices, _ = initialize_device_settings(use_cuda=use_gpu, multi_gpu=False)
+        device = 0 if devices[0].type == "cuda" else -1
+
+        self.model = pipeline(
+            task=task, model=model_name_or_path, tokenizer=tokenizer, device=device, revision=model_version
+        )
+
+        self.labels = labels
+        if task == "zero-shot-classification":
+            if labels is None or len(labels) == 0:
+                raise ValueError("Candidate labels must be provided for task zero-shot-classification")
+        elif task == "text-classification":
+            labels_from_model = [label for label in self.model.model.config.id2label.values()]
+            if labels is None:
+                self.labels = labels_from_model
+            elif set(labels) != set(labels_from_model):
+                self.labels = labels_from_model
+                logger.warning(
+                    f"The provided labels do not match the model labels, then the model labels are used.\n"
+                    f"Provided labels: {labels}\n"
+                    f"Model labels: {labels_from_model}"
+                )
+        else:
+            raise ValueError(
+                f"Task not supported: {task}.\n"
+                f"Possible task values are: 'text-classification' or 'zero-shot-classification'"
+            )
+        self.task = task
         self.batch_size = batch_size
-        device = 0 if self.devices[0].type == "cuda" else -1
-
-        model = AutoModelForSequenceClassification.from_pretrained(model_name_or_path)
-        tokenizer = AutoTokenizer.from_pretrained(model_name_or_path)
 
-        self.query_classification_pipeline = TextClassificationPipeline(model=model, tokenizer=tokenizer, device=device)
-
-    def run(self, query):
-        is_question: bool = self.query_classification_pipeline(query)[0]["label"] == "LABEL_1"
-
-        if is_question:
-            return {}, "output_1"
-        else:
-            return {}, "output_2"
+    @classmethod
+    def _calculate_outgoing_edges(cls, component_params: Dict[str, Any]) -> int:
+        labels = component_params.get("labels", None)
+        if labels is None:
+            return 2
```

Comment:
I'm a bit wary of this two lines. Using 2 as a fallback in case the actual number of edges is unknown can lead to tricky bugs, because it lies to the validation code. On the other hand, I understand that at this stage the node has no way to know how many labels there will be in the model without loading the model itself.

I see two ways to fix this:

- Heavyweight: try to extract the information from the model without loading it. Pro: users don't need to know which and how many labels the model supports. Cons: I foresee meddling with transformers internals to extract this info, which is never a good idea, or loading the model, which sounds overkill.

- Lightweight: remove the capability of this node to figure out which labels the model supports, and make `labels` mandatory. Pro: validation is now possible and correct. Cons: the users need to figure out the labels of the model on their own.

I prefer the lightweight because it's way more explicit, even if it forces the users to do some more legwork. WDYT?

Reply:
The problem with the current behavior is that it works if you're loading a pipeline from code, but not from YAML. Try to load a model with three labels from a YAML pipeline: it will fail, because the validation occurs before the init, and `_calculate_outgoing_edges` return 2. 

The old behavior works slightly better, but just because we're "brute-forcing" the problem. Given that most classifiers have about 2-3 labels, setting 10 as the fallback value reduces the number of times it will fail, but it also add bugs on the YAML side: now, if you model has two labels and you connect three nodes to it, the validation will pass, but the Pipeline will crash at load. Not too nice either.

I'm not 100% against this approach actually (that's how it was done before, and more or less it worked), but I wanted to make sure there is really no other way to go. I'll tag in the PR author (@tstadel), maybe he has some other idea?

Personally I like the lightweight approach because most users must know the order of labels anyway if they want to connect the right nodes to the right edge. They have to know whether the first edge is positive, negative or neutral, and to do so they must also find out which are the model labels. So adding them to the input parameters should not be an overhead at all. But if you don't like this, we can continue with the "default to 10 approach" for now. Let me know :slightly_smiling_face:

Reply:
>Personally I like the lightweight approach because most users must know the order of labels anyway if they want to connect the right nodes to the right edge. They have to know whether the first edge is positive, negative or neutral, and to do so they must also find out which are the model labels. So adding them to the input parameters should not be an overhead at all.
 
Now I understand your point of view better and it makes sense.
In fact, if users choose a custom model, they need to understand some details about it.
I will follow this approach!